test/unit/mongo_logger.test.ts

```diff
@@ -1548,4 +1549,70 @@ describe('class MongoLogger', async function () {
       });
     });
   });
+
+  describe('willLog', function () {
+    const expectedWillLogResults = {};
+    const errorWillLogRow = {
+      off: true,
+      emergency: true,
+      alert: true,
+      critical: true,
+      error: true,
+      warn: false,
+      notice: false,
+      info: false,
+      debug: false,
+      trace: false
+    };
+    const debugWillLogRow = {
+      off: true,
+      emergency: true,
+      alert: true,
+      critical: true,
+      error: true,
+      warn: true,
+      notice: true,
+      info: true,
+      debug: true,
+      trace: false
+    };
+
+    let logger: MongoLogger;
+
+    for (const component of Object.values(MongoLoggableComponent)) {
+      context(`when component is ${component}`, function () {
+        beforeEach(function () {
+          // create expectedWillLogResults
+          for (const component2 of Object.values(MongoLoggableComponent)) {
+            expectedWillLogResults[component2] =
+              component2 === component ? debugWillLogRow : errorWillLogRow;
+          }
+
+          // generate component severities, set default severity to 'error' and relevant component severity to 'debug'
+          const componentSeverities = Object.fromEntries(
+            Object.entries(MongoLoggableComponent).map(([_key, value]) =>
+              value === component ? [value, 'debug'] : [value, 'error']
+            )
+          );
+
+          // create logger with relevant component severities
+          logger = new MongoLogger({
+            componentSeverities: componentSeverities,
+            logDestination: createStdioLogger(process.stderr),
+            logDestinationIsStdErr: true
+          } as any);
+        });
+
+        it(`willLog method should only return true for values <= its componentSeverity`, function () {
+          for (const component3 of Object.values(MongoLoggableComponent)) {
+            for (const severityLevel of Object.values(SeverityLevel)) {
+              expect(logger.willLog(component3, severityLevel)).to.equal(
+                expectedWillLogResults[component3][severityLevel]
+              );
+            }
+          }
+        });
+      });
+    }
+  });
```

Comment:
```suggestion
  describe('#willLog', function () {
    const severityLevels = Object.values(SeverityLevel);
    for (const severityLevel of severityLevels) {
      context(`when the severity level is ${severityLevel}`, function () {
        let logger: MongoLogger;
        let componentSeverities;
        const components = Object.values(MongoLoggableComponent);

        for (const component of components) {
          context(`when ${component} severity level <= ${severityLevel}`, function () {
            beforeEach(function () {
              const index = severityLevels.indexOf(severityLevel);
              componentSeverities = components.reduce((severities, value) => {
                component === value
                  ? (severities[component] = severityLevel)
                  : (severities[value] =
                      severityLevels[index + 1] === 'off'
                        ? severityLevel
                        : severityLevels[index + 1]);
                return severities;
              }, {});
              logger = new MongoLogger({
                componentSeverities: componentSeverities,
                logDestination: createStdioLogger(process.stderr),
                logDestinationIsStdErr: true
              } as any);
            });

            if (severityLevel === 'off') {
              it('returns false always for off', function () {
                expect(logger.willLog(component, severityLevel)).to.be.false;
              });
            } else {
              it('returns true', function () {
                expect(logger.willLog(component, severityLevel)).to.be.true;
              });
            }
          });

          context(`when ${component} severity level > ${severityLevel}`, function () {
            if (severityLevel !== 'emergency') {
              beforeEach(function () {
                const index = severityLevels.indexOf(severityLevel);
                componentSeverities = components.reduce((severities, value) => {
                  component === value
                    ? (severities[component] = severityLevels[index - 1])
                    : (severities[value] = severityLevel);
                  return severities;
                }, {});
                logger = new MongoLogger({
                  componentSeverities: componentSeverities,
                  logDestination: createStdioLogger(process.stderr),
                  logDestinationIsStdErr: true
                } as any);
              });

              it('returns false', function () {
                expect(logger.willLog(component, severityLevel)).to.be.false;
              });
            }
          });
        }
      });
    }
  });
});

```

Ok so I spent some time going over these tests, and I'll explain the rational behind this suggestions.

First, the previous tests had one `it` block with multiple assertions. When looking at the test output it would not be clear on which component or severity level failed, so I refactored the tests to have an assertion per `it` block per component and severity level, thus giving us a test output like so:

```
class MongoLogger
    #willLog
      when the severity level is emergency
        when command severity level <= emergency
          ✔ returns true
        when topology severity level <= emergency
          ✔ returns true
        when serverSelection severity level <= emergency
          ✔ returns true
        when connection severity level <= emergency
          ✔ returns true
        when client severity level <= emergency
          ✔ returns true
      when the severity level is alert
        when command severity level <= alert
          ✔ returns true
        when command severity level > alert
          ✔ returns false
        when topology severity level <= alert
          ✔ returns true
        when topology severity level > alert
          ✔ returns false
        when serverSelection severity level <= alert
          ✔ returns true
        when serverSelection severity level > alert
          ✔ returns false
        when connection severity level <= alert
          ✔ returns true
        when connection severity level > alert
          ✔ returns false
        when client severity level <= alert
          ✔ returns true
        when client severity level > alert
          ✔ returns false
      when the severity level is critical
        when command severity level <= critical
          ✔ returns true
        when command severity level > critical
          ✔ returns false
        when topology severity level <= critical
          ✔ returns true
        when topology severity level > critical
          ✔ returns false
        when serverSelection severity level <= critical
          ✔ returns true
        when serverSelection severity level > critical
          ✔ returns false
        when connection severity level <= critical
          ✔ returns true
        when connection severity level > critical
          ✔ returns false
        when client severity level <= critical
          ✔ returns true
        when client severity level > critical
          ✔ returns false
      when the severity level is error
        when command severity level <= error
          ✔ returns true
        when command severity level > error
          ✔ returns false
        when topology severity level <= error
          ✔ returns true
        when topology severity level > error
          ✔ returns false
        when serverSelection severity level <= error
          ✔ returns true
        when serverSelection severity level > error
          ✔ returns false
        when connection severity level <= error
          ✔ returns true
        when connection severity level > error
          ✔ returns false
        when client severity level <= error
          ✔ returns true
        when client severity level > error
          ✔ returns false
      when the severity level is warn
        when command severity level <= warn
          ✔ returns true
        when command severity level > warn
          ✔ returns false
        when topology severity level <= warn
          ✔ returns true
        when topology severity level > warn
          ✔ returns false
        when serverSelection severity level <= warn
          ✔ returns true
        when serverSelection severity level > warn
          ✔ returns false
        when connection severity level <= warn
          ✔ returns true
        when connection severity level > warn
          ✔ returns false
        when client severity level <= warn
          ✔ returns true
        when client severity level > warn
          ✔ returns false
      when the severity level is notice
        when command severity level <= notice
          ✔ returns true
        when command severity level > notice
          ✔ returns false
        when topology severity level <= notice
          ✔ returns true
        when topology severity level > notice
          ✔ returns false
        when serverSelection severity level <= notice
          ✔ returns true
        when serverSelection severity level > notice
          ✔ returns false
        when connection severity level <= notice
          ✔ returns true
        when connection severity level > notice
          ✔ returns false
        when client severity level <= notice
          ✔ returns true
        when client severity level > notice
          ✔ returns false
      when the severity level is info
        when command severity level <= info
          ✔ returns true
        when command severity level > info
          ✔ returns false
        when topology severity level <= info
          ✔ returns true
        when topology severity level > info
          ✔ returns false
        when serverSelection severity level <= info
          ✔ returns true
        when serverSelection severity level > info
          ✔ returns false
        when connection severity level <= info
          ✔ returns true
        when connection severity level > info
          ✔ returns false
        when client severity level <= info
          ✔ returns true
        when client severity level > info
          ✔ returns false
      when the severity level is debug
        when command severity level <= debug
          ✔ returns true
        when command severity level > debug
          ✔ returns false
        when topology severity level <= debug
          ✔ returns true
        when topology severity level > debug
          ✔ returns false
        when serverSelection severity level <= debug
          ✔ returns true
        when serverSelection severity level > debug
          ✔ returns false
        when connection severity level <= debug
          ✔ returns true
        when connection severity level > debug
          ✔ returns false
        when client severity level <= debug
          ✔ returns true
        when client severity level > debug
          ✔ returns false
      when the severity level is trace
        when command severity level <= trace
          ✔ returns true
        when command severity level > trace
          ✔ returns false
        when topology severity level <= trace
          ✔ returns true
        when topology severity level > trace
          ✔ returns false
        when serverSelection severity level <= trace
          ✔ returns true
        when serverSelection severity level > trace
          ✔ returns false
        when connection severity level <= trace
          ✔ returns true
        when connection severity level > trace
          ✔ returns false
        when client severity level <= trace
          ✔ returns true
        when client severity level > trace
          ✔ returns false
      when the severity level is off
        when command severity level <= off
          ✔ returns false always for off
        when command severity level > off
          ✔ returns false
        when topology severity level <= off
          ✔ returns false always for off
        when topology severity level > off
          ✔ returns false
        when serverSelection severity level <= off
          ✔ returns false always for off
        when serverSelection severity level > off
          ✔ returns false
        when connection severity level <= off
          ✔ returns false always for off
        when connection severity level > off
          ✔ returns false
        when client severity level <= off
          ✔ returns false always for off
        when client severity level > off
          ✔ returns false


  95 passing (71ms)

wrote xunit.xml                                                                                                                          
```

This gives us much better insight when looking at build failure logs as to what has failed specifically in these tests, and makes it easier to reason about what exactly the problem is before investigation.

Secondly, in my refactor I changed the setup to handle both sides of the comparison, <= and >. Setting up the logger options to use the indexes of the security level enum to dynamically generate test that cover every possible combination. This lead to...

Thirdly, the generation of the severities object was not quite correct for the way in which it was accessed, and the test refactor exposed a bug in the handling of the "off" severity. In the creation of the pre-generated object, "off" would be set to `true` if was set in the options for a component, causing `willLog` to return `true` if "off" was set on the component. This would have been the same outcome in the previous iterations of this PR. You will see in my new suggestion for the `willLog` method will address that. I do note that "off" would never be exposed via the public logging methods we use as it's not bound as a logging method, but this will future proof possible issues.

Apologies for the huge section here. I can if you like push to this branch with my local changes posted here, or feel free to change yourself. Great work so far BTW.